Add hooks, counters and refine the statistics infrastructure

jvican opened the pull request:
This is a continuation of my work for SCP-010. This pull request is the responsible of polishing the statistics framework, adding missing hooks for our Scala plugin scalac-profiling and double-checking the general state of statistics.

As a result, this pull request ends up doing several independent tasks:
1. It instruments scopes so that we can analyze them later on.
1. It adds timers for all the phases (`-Ystatistics` will report how much time do phases take).
1. Adds hooks to `AnalyzerPlugins` to inspect implicit search data and therefore allow external plugins to turn it into something meaningful.
1. Polishes some parts of the user interface of statistics (flags, descriptions).
1. Fixes/cleans some parts of the infrastructure that is unstable or buggy.

Depends on #6051 and #6034.

jvican committed:
Print timers of phases at the end

The following commit adds the ability to print the timers for every
phase at the end of the `compileSources` execution. This is useful
because you can have an idea of how the running times of the phases are
distributed. It also works for phases injected by the incremental
compiler and compiler plugins (which allows devs to estimate how much
time do their compiler plugins take out of the total compile time).

It also removes the previous infrastructure to print these timings under
the verbose flag, and now reuses the full statistics infrastructure for
doing so.

jvican committed:
Add statistics to scopes handling

This commit adds two important things to the scopes handling in scalac:

* Count how many scopes are created.
* Time common scope operations, like population and lookup so that we
  can effectively measure the impact of unused imports and the like.

jvican committed:
Add padding to implicits timers

lrytz commented on src/reflect/scala/reflect/internal/Scopes.scala:
```diff
@@ -8,9 +8,12 @@ package reflect
 
 trait Scopes extends api.Scopes { self: SymbolTable =>
 
+  private[scala] var scopeCount = 0
```
Should this be reset to 0 after each run? That could be achieved with `perRunCaches.newGeneric`.

jvican replied:
Hm, yes, it should be reset to 0 after every run. I wanted to emulate trees ids here, that's why I used just one counter.

jvican replied:
I didn't figure out how to do it with `newGeneric`, so I implemented with a clearable. See updated PR. Thanks for the suggestion.

lrytz replied:
OK, `newGeneric` was probably not a good suggestion here

lrytz commented on src/compiler/scala/tools/nsc/Global.scala:
```diff
@@ -1431,26 +1431,31 @@ class Global(var currentSettings: Settings, var reporter: Reporter)
       while (globalPhase.hasNext && !reporter.hasErrors) {
-        val startTime = currentTime
         phase = globalPhase
+        val phaseTimer = statistics.newSubTimer(s"  ${phase.name}", totalCompileTime)
```
Can we make this `null` if `! timePhases`?

lrytz commented on src/compiler/scala/tools/nsc/Global.scala:
```diff
@@ -1431,26 +1431,31 @@ class Global(var currentSettings: Settings, var reporter: Reporter)
 
         // progress update
-        informTime(globalPhase.description, startTime)
+        informTime(globalPhase.description, phaseTimer.nanos)
```
should this be guarded by `if (timePhases)`?

lrytz commented on src/compiler/scala/tools/nsc/Global.scala:
```diff
@@ -1503,7 +1508,13 @@ class Global(var currentSettings: Settings, var reporter: Reporter)
+        statistics.stopTimer(totalCompileTime, startTotal)
+        informTime("total", totalCompileTime.nanos)
+        inform("*** Cumulative timers for phases")
+        for (q <- statistics.allQuantities if q.phases.nonEmpty && q.showAt(GlobalPhaseName))
```
Nit: the guard `q.phases.nonEmpty && q.showAt(GlobalPhaseName)` is non-obvious to understand - how about the more explicit `q.phases == Seq(GlobalPhaseName)`?

jvican replied:
We can use this because `GlobalPhaseName` will always be the only phase, but it's more restrictive semantically.

lrytz commented on src/compiler/scala/tools/nsc/typechecker/Implicits.scala:
```diff
@@ -1578,6 +1581,7 @@ trait ImplicitsStats {
   val matchingImplicits   = newSubCounter("  #matching", implicitSearchCount)
   val typedImplicits      = newSubCounter("  #typed", implicitSearchCount)
   val foundImplicits      = newSubCounter("  #found", implicitSearchCount)
+  val macroImplicits      = newSubCounter("  #macros", implicitSearchCount)
```
Can we give it a more telling name, e.g. `#triggered during macro expansion`?

lrytz replied:
This was not addressed in your latest update.

lrytz commented on src/compiler/scala/tools/nsc/typechecker/Implicits.scala:
```diff
@@ -1581,4 +1591,8 @@ trait ImplicitsStats {
+
+  import scala.reflect.internal.util.Position
+  final val implicitSearchesByType = symbolTable.perRunCaches.newMap[global.Type, Int]()
+  final val implicitSearchesByPos = symbolTable.perRunCaches.newMap[Position, Int]()
```
Instead of collecting this data within the compiler (this data is collected when ever statistics are enabled, but only used by your compiler plugin), I'd prefer some mechanism to allow a compiler plugin to listen for certain events. See [`AnalyzerPlugins`](https://github.com/scala/scala/blob/v2.12.3/src/compiler/scala/tools/nsc/typechecker/AnalyzerPlugins.scala#L281-L288) for an example. Here it would be simpler than with analyzer plugins, because it's only about listening for notifications and not changing the result of any method call.

I'm open for other suggestions.

lrytz replied:
This was not addressed. The goal here is to provide a hook for compiler plugins, so I really feel that it's not great to collect some very specific data inside the compiler itself.

jvican replied:
Hey Lukas, I don't have a clear idea how we could engineer a general interface to listen for general events in the compiler that adds more value than this current implementation.
  
The problem I see here is that these hooks are very explicit and require instrumentation per ever kind of event (in different parts of the compiler).
  
In `AnalyzerPlugins`, I see a more general interface that allows plugins to refine type signatures in certain places, giving a feeling that analyzer plugins can complement typechecking. I don't see a way we can do this for explicit event listening because this requires:
  
* The creation of the timers and subtimers.
* Exposing these timers and subtimers explicitly in the `statistics` field, so that they are accessible from any call-site of the compiler.
* Instrumenting one call-site per ever event you want to listen.
  
Maybe I'm missing something?

retronym replied:
I think that implicit search should be instrumented here in the compiler to keep track of some important statistics and aggregate these into `SearchResult`. The most important stat is the overall number of `typedImplicit1` calls made within an implicit search tree. Perhaps the number of implicit macros expanded is also interesting. We should also have an easy way for the tools which process the data to distinguish "top level" implicit searches from recursive ones.

A new extension point in `AnalyzerPlugin`-s could be added to listen on the result of `inferImplicit`, and then the work to slice, dice, and present this data could be left in the hands of the plugin.

lrytz replied:
Thanks for the input. If this PR can still make it for 2.12.4 I'm happy to have a minimal solution here for now and implement a more structured approach later. Would that work for you, @retronym? Would you be OK with the two hash maps?

jvican replied:
@retronym @lrytz This sounds like a good idea. I'm implementing it real quick. However, I'll add two hooks into implicit search, one before the search and the other after the search. This will help me add timers and not only counters.

jvican replied:
Implemented in https://github.com/scala/scala/pull/6067/commits/41d60f836be54a7010c3656e5e28107451bfea54.

lrytz commented on src/reflect/scala/reflect/internal/Scopes.scala:
```diff
@@ -8,12 +8,21 @@ package reflect
   private[scala] var scopeCount = 0
 
+  // Reset scopeCount per every run
+  perRunCaches.recordCache(new Clearable {
```
you could use a lambda, `Clearable` is a SAM type

jvican replied:
I tried -- it didn't work because the compiler failed to create a SAM type because `recordCache`'s signature expects `T <: Clearable`. It seems SAM creation doesn't work for these cases (or so I think).

lrytz replied:
ah right, you'd need a type annotation `(() => { scopeCount = 0 }): Clearable`

jvican replied:
What I ended up doing locally is creating a new private val with the closure. IMO it's better than the type annotation. If you don't have any objection, I'll continue this way.

lrytz replied:
Either way is fine. A local val would be better than a private val in `Scopes`, as the latter creates a field.

lrytz reviewed (commented):
For some reason two of my earlier comments showed up collapsed on the PR discussion even though they were up to date. Maybe a bug in GitHub.
  - https://github.com/scala/scala/pull/6067/commits/69ef9a50671544b501dd5393fe8b83e42b808622#r140470518
  - https://github.com/scala/scala/pull/6067/commits/5e57b299f721cc42303eabbd9fb2a7f00502d24d#r140473690

jvican commented:
Indeed, it seems I missed completely these two comments.

lrytz commented:
Some of the commits have the following test failure:
```
!! 1278 - pos/t8531                                 [compilation failed]
##### Log file '/home/jenkins/workspace/scala-2.12.x-validate-test/test/files/pos/t8531-pos.log' from failed test #####

error: null
```

The fact that it happens on some commits but not on all suggests it could be a race condition (AFAIK, partest runs multiple `Global` instances in parallel).

adriaanm commented:
Looks more likely this will slip to 2.12.5 then.

jvican commented:
@lrytz I would be really surprised if this had any thread safety issues as it just adds some counters and timers. If this code had any thread safety issues, all the statistics infrastructure would, which I find pretty unlikely given the code and my experience using/hacking on it.
  
I think the problem here is that some commits were rebased and there were some issues with this rebase. Let me look into this and quickly push the new changes.
  
> Looks more likely this will slip to 2.12.5 then.
  
I would really love if this can be merged after my changes have been done and all commits are green. This change is non controversial -- it's not like my other PR that makes a fundamental change to the infrastructure.

retronym commented on src/reflect/scala/reflect/internal/Scopes.scala:
```diff
@@ -229,36 +239,41 @@ trait Scopes extends api.Scopes { self: SymbolTable =>
     /** Lookup a symbol.
      */
     def lookup(name: Name): Symbol = {
+      val startTime = if (statistics.canEnable) statistics.startTimer(statistics.scopeLookupTime) else null
```
I don't think this is really a suitable place to instrument, as it misses a lot of important lookups that use alternative scope lookup APIs. E.g. `Context.isQualifyingImplicit` uses `lookupEntry` and `Context.lookupSymbol` uses `lookupUnshadowedEntries`.

retronym replied:
Oh, I see you've instrumented `lookupEntry` as well below. Should we then start this timer below the call to `lookupEntry` to avoid double counting?

jvican replied:
I instrumented both so that `lookupEntry` also records the time spent in executing its whole body, which relies on `lookupNextEntry`. I think that the statistics infrastructure prevents double counting by design (nested calls do not record the times if there are other timers wrapping them) -> a fresh pair of eyes to confirm this would be good.
  
However, maybe measuring `lookup` is uninteresting... so I've removed it in my last commit.

lrytz commented:
I didn't spot any issues from rebasing when I looked at the commits. The error message in the test failure is as unhelpful as it gets (`error: null` without *any* context), that's really annoying.

`t8531` is about the pattern matcher / exhaustivity analysis, but I'm not sure that is a hint in the right direction.

retronym commented:
> that's really annoying.

I can get that deficient error message by throwing an NPE in the compiler, or any exception with a null message.

I'm guessing that partest is culpable for the old favourite, [`println(throwable.getMessage)`](https://github.com/scala/scala-partest/blob/4d2b327b48da98553f8efd2e3bbc55052d68642d/src/main/scala/scala/tools/partest/nest/DirectCompiler.scala#L124). But I'm not 100% sure, we've got similarly horrifying exception handling in the [compiler itself](https://github.com/scala/scala/blob/a74f0c7a1457b8a17db36db25d068dc1583c575e/src/compiler/scala/tools/nsc/Global.scala#L1508).

A quick fix to diagnose this would be to add a commit that changes `Run.compileInternal` to catch `Throwable` and print the results to system out. That should show up in the partest log.

jvican committed:
Allow `AnalyzerPlugin`s to hook into implicit search

The following commit allows an external analyzer plugin to hook into
scalac's implicit search. This change replaces explicit instrumentation
of this part to capture statistics, and has been therefore extended to
allow more generic scenarios, leaving the tasks of capturing the data to
external third parties (via compiler plugins).

The change adds two new members to the public API of `AnalyzerPlugin`:

1. `pluginsImplicitSearch`: `ImplicitSearch => ()`.
1. `pluginsImplicitSearchResult`: `SearchResult` => `()`.

`ImplicitSearch` is the data structure that contains all the necessary
information to perform implicit search, whereas `SearchResult` is the
resulting found implicit instance.

These two methods allow the analyzer plugin to access the full context
of implicit search. In order to have access to the parameters of
`ImplicitSearch`, this commit also makes private parameters accessible
by converting them to `val`s.

jvican commented:
@lrytz @retronym Can you have a look at the new `AnalyzerPlugin` hooks?
  
Everything is green and ready for a review.

jvican commented:
Benchmark scheduled [here](https://scala-ci.typesafe.com/grafana/dashboard/db/pr-validation?var-scalaSha=8a92ebdce5|1be1eb7bee&var-source=All&var-bench=HotScalacBenchmark.compile&var-host=scalabench@scalabench@&from=now-30d&to=now).

adriaanm commented:
We'll keep the queue open until Friday, run benchmarks and community builds over the weekend and Monday, and cut the release on Tuesday. Any PRs that regress performance / break the community build will be reverted and rescheduled for 2.12.5.

jvican commented:
@adriaanm Sounds good. I'll run the community build before you do to make sure all my PRs are ok.

retronym commented:
@jvican Did you figure out the root cause of that null-messaged exception? If not, I'm a bit worried that either this PR or something recent has introduced an non-deterministic error that we need to track down before 2.12.4.

retronym commented on src/compiler/scala/tools/nsc/typechecker/Implicits.scala:
```diff
@@ -90,7 +90,9 @@ trait Implicits {
-    val result = new ImplicitSearch(tree, pt, isView, implicitSearchContext, pos).bestImplicit
+    val search0 = new ImplicitSearch(tree, pt, isView, implicitSearchContext, pos)
+    val search = pluginsImplicitSearch(search0)
+    val result = pluginsImplicitSearchResult(search.bestImplicit)
```
I'm not really comfortable letting plugins change the results of an implicit search. A less powerful extension point would just have `pluginsImplicitSearchResult(search: ImplicitSearch): Unit`.

The standard `ImplicitSearch` can itself contain a timer, which is conditionally computed if stats are on. We don't need to leave that part to the plugin.

jvican replied:
> I'm not really comfortable letting plugins change the results of an implicit search. A less powerful extension point would just have pluginsImplicitSearchResult(search: ImplicitSearch): Unit.
  
I've adapted it to be a simple hook. `AnalyzerPlugins` is itself a powerful tool that refines types from the typechecker, so I wasn't sure that such a powerful extension point would not be accepted since it aligns with the overall idea of the plugin.
  
> The standard ImplicitSearch can itself contain a timer, which is conditionally computed if stats are on. We don't need to leave that part to the plugin.
  
We already have a timer for this, e.g. `implicitNanos`, and I can access it statically in my compiler plugin, so I think this is not necessary.
  
However, I didn't realize that we have another `ImplicitSearch` for views. They should mostly be unused, but just in case I've added a timer for them too.

lrytz commented on src/compiler/scala/tools/nsc/typechecker/AnalyzerPlugins.scala:
```diff
@@ -349,6 +369,18 @@ trait AnalyzerPlugins { self: Analyzer =>
   })
 
+  /** @see AnalyzerPlugin.pluginsImplicitSearch */
+  def pluginsImplicitSearch(search: ImplicitSearch): ImplicitSearch = invoke(new CumulativeOp[Unit] {
```
Should have return type `Unit` (also below)

jvican replied:
Yes, I pushed a bad commit.

lrytz commented on src/compiler/scala/tools/nsc/typechecker/Implicits.scala:
```diff
@@ -1585,6 +1590,7 @@ trait ImplicitsStats {
   val improvesCachedCount = newSubCounter("  #implicit improves cached ", implicitSearchCount)
   val inscopeImplicitHits = newSubCounter("  #implicit inscope hits", implicitSearchCount)
   val oftypeImplicitHits  = newSubCounter("  #implicit oftype hits ", implicitSearchCount)
+  val viewsNanos          = newSubTimer  ("time spent in views", typerNanos)
```
I think this description is a bit misleading: `allViewsFrom` seems to be used only by scaladoc, the compiler uses `inferImplicit` with `isView = true`.

jvican replied:
Intersting, didn't know this. Then I'm removing it.

lrytz reviewed (commented):
It looks good to me now. I'm also still a bit worried about the null-message. But looking through the commits I really cannot see anything suspicious.

jvican committed:
Hot fix registration of timers and counters

Because of initialization order of the cake, `statistics` is initialized
after the timers/counters in the cake are, so when it hits the
constructor of `Quantity` those are not registered in `qs`.

This meant that even though those objects were initialized, statistics
were not reported. This change hot fixes it so that they are indeed
reported. It does so by removing the guard that checked whether
statistics were enabled. From now on, for simplicity and correctness, we
will always register timers and counters that are initialized. This
should have no impact in performance, since it's done only once when
everything is initialized, and it's just an addition to a map.

jvican commented:
> It looks good to me now. I'm also still a bit worried about the null-message. But looking through the commits I really cannot see anything suspicious.
  
Me neither... I've gone through my previous PRs too and didn't find anything suspicious re thread safety. I'll go through them again, just in case there's something hidden.
  
I'm integrating all the changes with my scala-profiling integration, and double-checking that everything works. I'll be running the community build with only my changes to check that there's no issue. In a few hours, I'll have some results to show.

jvican committed:
Don't display hot counters if hot statistics are disabled

This is purely a cosmetic change to make sure that these counters, which
will always be zero or empty, are not displayed to users in case hot
statistics are disabled.

jvican committed:
Change name and description of `-Yhot-statistics-enabled`

This is a pure UI change that makes it clear that
`-Yhot-statistics-enabled` is a complement of `-Ystatistics`, not a
replacement. Therefore, users need to use it together with
`-Ystatistics`, otherwise `-Yhot-statistics-enabled` won't do anything.

jvican commented:
Community build is still running on my cluster, but so far I'm getting no errors.

jvican committed:
Avoid disabling statistics

This commit avoids disabling statistics to avoid the JVM to unstabilize
all the time and to prevent misbehaviours in concurrent builds recycling
the same classloader.

The main problem here is that if one global has statistics enabled and
the second one doesn't, our logic of `enabled` was setting the
pseudo-static to false, preventing the correct recording of statistics
in the global that does want to record them. Now, the logic to disable
these pseudo-statics when `statistics = false` (which is the case almost
always), has been removed.

TL;DR: Once a global enables statistics, the rest has to pay the price
of a 2/3% performance hit. There is no going back.

jvican commented:
Today I've spent almost all day double-checking this PR and [integrating it with scalac-profiling](https://github.com/scalacenter/scalac-profiling/pull/7). This process has given me confidence that I have everything I need from the compiler after the exhaustive feedback that I received in this and the previous PRs.
  
These are the tasks I've done to double-check this PR is in good shape:
  
* Compile both the library and the compiler with the binaries and the following flags: `-Ystatistics`, `-Yhot-statistics-enabled`. Everything ok.
* Enabled `-Ystatistics` in one project in an sbt multi-build and double-checked that only that project outputs statistics (the other one doesn't, despite the fact that they share the same classloader). Everythink ok.
* Executed different projects with statistics enabled in parallel. Everything ok.
* Run the community build in our cluster (it's still running but so far everything passes).
  
As a result of this process, I've updated this PR with minor fixes to the UI and the semantics of the whole statistics framework. I'm fairly confident this PR doesn't represent a problem for the upcoming release.
  
I would appreciate if someone can go through it again and tell me if it looks good to you.
  
Ping me during the weekend if you have any question regarding my PRs.

adriaanm commented:
Thanks for the due diligence, @jvican. I'll let @retronym take a look his Monday morning and hopefully get this merged by my Sunday night :-)

lrytz commented on src/compiler/scala/tools/nsc/settings/ScalaSettings.scala:
```diff
@@ -384,7 +384,7 @@ trait ScalaSettings extends AbsScalaSettings
   override def YstatisticsEnabled = Ystatistics.value.nonEmpty
 
-  val YhotStatistics = BooleanSetting("-Yhot-statistics", "Print hot compiler statistics for all relevant phases")
+  val YhotStatistics = BooleanSetting("-Yhot-statistics-enabled", s"Enable `${Ystatistics.name}` to print hot statistics.")
```
You could make `-Ystatistics` a `ChoiceSetting` (`-Ystatistics:{none|cold|all}`), but that's not a blocker for merging this PR.

rorygraves commented:
Woot!

> On 2 Oct 2017, at 00:01, Adriaan Moors <notifications@github.com> wrote:
> 
> Merged #6067.
> 
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub, or mute the thread.
>
